### PR TITLE
refactor: explicitly assign every internal config field

### PR DESCRIPTION
**Pull request by kjin**
Prerequisite to addressing a feature request in #994... this change makes every internal config option assignment explicit, and re-organizes internal config options so that instead of `interface ParentConfig extends ChildConfig` it becomes `interface ParentConfig { childConfig: ChildConfig }`. This means that each configurable component (internally) only gets config options that it asks for, rather than a large object containing _all_ config options.

Tested behavior should remain the same; some tests have been changed because of the re-organization of internal options.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -35,73 +36,132 @@ let traceAgent: StackdriverTracer;
...
+  return {
+    [FORCE_NEW]: forceNew,
+    enabled: projectConfig.enabled,
+    logLevel: lastOf(
+        projectConfig.logLevel, Number(process.env.GCLOUD_TRACE_LOGLEVEL)),
+    clsConfig: {
+      [FORCE_NEW]: forceNew,
+      mechanism: transform(
+          projectConfig.clsMechanism,
+          (clsMechanism):
```
Comment (ofrobots) [inline src/index.ts]: Too much details inline. Can you extract this out of line in a named function.

File: src/index.ts [hunk 2]
```diff
@@ -35,73 +36,132 @@ let traceAgent: StackdriverTracer;
...
-
-  // If the CLS mechanism is set to auto-determined, decide now what it should
-  // be.
-  const ahAvailable = semver.satisfies(process.version, '>=8');
-  if (config.clsMechanism === 'auto') {
-    config.clsMechanism = ahAvailable ? 'async-hooks' : 'async-listener';
-  }
-
-  return config;
+  const projectConfig: (typeof defaultConfig)&Forceable<Config> =
```
Comment (ofrobots) [inline src/index.ts]: `mergedConfig` would be a better name.

File: src/index.ts [hunk 3]
```diff
@@ -35,73 +36,132 @@ let traceAgent: StackdriverTracer;
...
+            projectConfig.serviceContext.version,
+            process.env.GAE_MODULE_VERSION, process.env.GAE_VERSION),
+        minorVersion: lastOf<string|undefined>(
+            projectConfig.serviceContext.minorVersion,
+            process.env.GAE_MINOR_VERSION)
+      }
+    },
+    pluginLoaderConfig: {
+      [FORCE_NEW]: forceNew,
+      plugins: Object.assign({}, projectConfig.plugins),
```
Comment (ofrobots) [inline src/index.ts]: style nit: `plugins: {... projectConfig.plugins)`

File: src/index.ts (hunk 2 above)
Comment (kjin) [reply src/index.ts]: Done

File: src/index.ts (hunk 1 above)
Comment (kjin) [reply src/index.ts]: Done

File: src/index.ts (hunk 3 above)
Comment (kjin) [reply src/index.ts]: IIRC this is not available in Node 6... which we haven't officially ripped out of CI yet.

Comment (ofrobots) [reply src/index.ts]: it is available in TypeScript tho

Comment (kjin) [reply src/index.ts]: We have to target es2015 to preserve generators, so it's not available even in TypeScript, unfortunately.

Comment (kjin) [reply src/index.ts]: Ok, that being said I'll probably just drop node6 CI in a different PR.

Comment (kjin) [reply src/index.ts]: Done

File: src/index.ts [hunk 4]
```diff
@@ -35,73 +36,128 @@ let traceAgent: StackdriverTracer;
...
-      minorVersion: process.env.GAE_MINOR_VERSION || undefined
-    }
-  };
-
-  let envSetConfig: Config = {};
+function initConfig(userConfig: Forceable<Config>): Forceable<TopLevelConfig> {
+  // Helper function which gets the last non-null/undefined value among its
+  // arguments. The first argument must therefore be guaranteed to be
+  // non-null/undefined.
+  const lastOf =
```
Comment (ofrobots) [inline src/index.ts]: this should be unit-tested.

File: src/index.ts [hunk 5]
```diff
@@ -35,73 +36,128 @@ let traceAgent: StackdriverTracer;
...
+      <T>(defaultValue: T, ...otherValues: Array<T|null|undefined>): T => {
+        for (let i = otherValues.length - 1; i >= 0; i--) {
+          // tslint:disable:no-any
+          if (otherValues[i] != null &&
```
Comment (ofrobots) [inline src/index.ts]: Is the `!=` intentional or did you want a `!==`?

File: src/index.ts [hunk 6]
```diff
@@ -35,73 +36,128 @@ let traceAgent: StackdriverTracer;
...
+    const ahAvailable = semver.satisfies(process.version, '>=8');
+    if (clsMechanism === 'auto') {
+      return ahAvailable ? TraceCLSMechanism.ASYNC_HOOKS :
+                           TraceCLSMechanism.ASYNC_LISTENER;
+    }
+    return clsMechanism as TraceCLSMechanism;
+  };
+  const getInternalMaximumLabelValueSize = (maximumLabelValueSize: number) => {
+    // Enforce the upper limit for the label value size.
+    if (maximumLabelValueSize > Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT) {
```
Comment (ofrobots) [inline src/index.ts]: nit: Would this work:

```ts
return Math.min(Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT, maximumLabelValueSize);
```

File: src/index.ts [hunk 7]
```diff
@@ -35,73 +36,128 @@ let traceAgent: StackdriverTracer;
...
+      return Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT;
+    }
+    return maximumLabelValueSize;
+  };
+  const getInternalRootSpanNameOverride =
+      (rootSpanNameOverride: string|((name: string) => string)) => {
+        // Make rootSpanNameOverride a function if not already.
```
Comment (ofrobots) [inline src/index.ts]: Style nit: is this more readable:
```ts
switch (typeof rootSpanNameOverride) {
  case 'string': return () => rootSpanNameOverride;
  case 'function': return (name: string) => name;
  default: return rootSpanNameOverride;
}
```

File: src/index.ts (hunk 5 above)
Comment (kjin) [reply src/index.ts]: It was intentional. That being said, I've separated it out into `!== undefined` and `!== null` for readability.

File: src/index.ts (hunk 7 above)
Comment (kjin) [reply src/index.ts]: Yeah that's better, thanks!

File: src/index.ts (hunk 4 above)
Comment (kjin) [reply src/index.ts]: Done, I've moved it into `src/util.ts`.

File: src/index.ts [hunk 8]
```diff
@@ -35,73 +36,109 @@ let traceAgent: StackdriverTracer;
...
+    // If the CLS mechanism is set to auto-determined, decide now
+    // what it should be.
+    const ahAvailable = semver.satisfies(process.version, '>=8');
+    if (clsMechanism === 'auto') {
+      return ahAvailable ? TraceCLSMechanism.ASYNC_HOOKS :
+                           TraceCLSMechanism.ASYNC_LISTENER;
+    }
+    return clsMechanism as TraceCLSMechanism;
+  };
+  const getInternalMaximumLabelValueSize = (maximumLabelValueSize: number) =>
```
Comment (ofrobots) [inline src/index.ts]: nit: now that this is a single line function, it would be okay to move it inline. I'll leave it up to you though.

**Pull request merged**